### PR TITLE
Improve conditions in block insert to determine winning obliterate

### PR DESCRIPTION
**Pull request by anthony-murphy**
Refactor how we determine if an obliterate applies primarily an insert. This added coverage for multiple local obliterates, and adds handling to make sure a local client can always insert over existing obliterates. All tests are passing locally.

### CODE REVIEW

**Comment by anthony-murphy** [comment]
i merged in the client farm changes which was hitting a couple issues, but made some updates to cover that, all tests are passing in the merge tree folder.

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 1]
```diff
@@ -1562,29 +1562,47 @@ export class MergeTree {
...
+							movedSeqs.push(ob.seq);
 						} else {
 							movedClientIds.push(ob.clientId);
 							movedSeqs.push(ob.seq);
 						}
 					}
 				}
 
-				if (found) {
+				if (oldest && newest?.clientId !== clientId) {
```
Comment (anthony-murphy) [inline packages/dds/merge-tree/src/mergeTree.ts:1589]: we use the newest obliterate to determine if the client can bypass the obliterate, but we use the data from the oldest obliterate which matches existing behavior

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 2]
```diff
@@ -1562,29 +1562,47 @@ export class MergeTree {
...
+							normalizedOldestSeq = normalizedObSeq;
+							oldest = ob;
 							movedClientIds.unshift(ob.clientId);
 							movedSeqs.unshift(ob.seq);
+						}
+						if (newest === undefined && normalizedNewestSeq < normalizedObSeq) {
+							normalizedNewestSeq = normalizedObSeq;
+							newest = ob;
+							movedClientIds.push(ob.clientId);
+							movedSeqs.push(ob.seq);
```
Comment (Abe27342) [inline packages/dds/merge-tree/src/mergeTree.ts]: the control flow here is a bit funky, can it be simplified by moving

```
movedClientIds.push(ob.clientId);
movedSeqs.push(ob.seq);
```
to outside all the nested if/else (i.e. whenever `normalizedObSeq > refSeq`, we note that this obliterate was one that applied to the segment, even if it wasn't necessarily newest/oldest)

File: packages/dds/merge-tree/src/mergeTree.ts [hunk 3]
```diff
@@ -1562,29 +1562,47 @@ export class MergeTree {
...
+							? Number.MAX_SAFE_INTEGER - this.collabWindow.localSeq + ob.localSeq!
+							: ob.seq;
+					if (normalizedObSeq > refSeq) {
+						if (oldest === undefined || normalizedOldestSeq > normalizedObSeq) {
+							normalizedOldestSeq = normalizedObSeq;
+							oldest = ob;
 							movedClientIds.unshift(ob.clientId);
 							movedSeqs.unshift(ob.seq);
+						}
+						if (newest === undefined && normalizedNewestSeq < normalizedObSeq) {
```
Comment (jzaffiro) [inline packages/dds/merge-tree/src/mergeTree.ts]:
```suggestion
						if (newest === undefined || normalizedNewestSeq < normalizedObSeq) {
```
This won't ever be true after the first obliterate in the overlapping array, which seems like a problem.

Comment (anthony-murphy) [reply packages/dds/merge-tree/src/mergeTree.ts]: nice catch! re-running the tests now.

**Review by jzaffiro — APPROVED** [review]

**Pull request merged**